url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Decoder.java

```diff
@@ -0,0 +1,192 @@
+//  of the protocol used. 0MQ framing protocol is based size-prefixed
+//  paradigm, which qualifies it to be parsed by this class.
+//  On the other hand, XML-based transports (like XMPP or SOAP) don't allow
+//  for knowing the size of data to read in advance and should use different
+//  decoding algorithms.
+//
+//  This class implements the state machine that parses the incoming buffer.
+//  Derived class should implement individual state machine actions.
+
+public class Decoder extends DecoderBase {
+    
+    private final static int one_byte_size_ready = 0;
+    private final static int eight_byte_size_ready = 1;
+    private final static int flags_ready = 2;
+    private final static int message_ready = 3;
+    
+    private final byte[] tmpbuf;
+    private Msg in_progress;
+    private final long maxmsgsize;
+    private IMsgSink msg_sink;
+
+    public Decoder (int bufsize_, long maxmsgsize_)
+    {
+        super(bufsize_);
+        maxmsgsize = maxmsgsize_;
+        tmpbuf = new byte[8];
+        
+    
+        //  At the beginning, read one byte and go to one_byte_size_ready state.
+        next_step (tmpbuf, 1, one_byte_size_ready);
+    }
+    
+
+    //  Set the receiver of decoded messages.
+// JDK6    @Override
+    public void set_msg_sink (IMsgSink msg_sink_) 
+    {
+        msg_sink = msg_sink_;
+    }
+
+    @Override
+    protected boolean next() {
+        switch(state()) {
+        case one_byte_size_ready:
+            return one_byte_size_ready ();
+        case eight_byte_size_ready:
+            return eight_byte_size_ready ();
+        case flags_ready:
+            return flags_ready ();
+        case message_ready:
+            return message_ready ();
+        default:
+            return false;
+        }
+    }
+
+
+
+    private boolean one_byte_size_ready() {
+        //  First byte of size is read. If it is 0xff(-1 for java byte) read 8-byte size.
+        //  Otherwise allocate the buffer for message data and read the
+        //  message data into it.
+        byte first = tmpbuf[0];
+        if (first == -1) {
+            next_step (tmpbuf, 8, eight_byte_size_ready);
+        } else {
+
+            //  There has to be at least one byte (the flags) in the message).
+            if (first == 0) {
+                decoding_error ();
+                return false;
+            }
+            
+            int size = (int) first;
+            if (size < 0) {
+                size = (0xFF) & first;
+            }
+
+            //  in_progress is initialised at this point so in theory we should
+            //  close it before calling zmq_msg_init_size, however, it's a 0-byte
+            //  message and thus we can treat it as uninitialised...
+            if (maxmsgsize >= 0 && (long) (size - 1) > maxmsgsize) {
+                decoding_error ();
+                return false;
+
+            }
+            else {
+                in_progress = new Msg(size-1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *in\_progress*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).